File: api_service/src/services/auth/auth_gateway_authority.py
from typing import Protocol

from src.models.auth.auth import (
    DownstreamAccessGrant,
    Session,
    TokenBundle,
    TokenVerificationResult,
)
from src.models.auth.auth_contract import (
    DownstreamGrantRequest,
    SessionValidateRequest,
    TokenRefreshRequest,
    TokenRevokeRequest,
    TokenVerifyRequest,
)
from src.models.auth.bootstrap import (
    BootstrapStage,
    BootstrapAuthRequest,
    BootstrapAuthResult,
    ChallengePayload,
    ChallengeRequest,
)


class IAuthGatewayAuthorityClient(Protocol):
    """认证中心统一鉴权门面客户端协议。"""

    async def init_challenge(self, req: ChallengeRequest) -> ChallengePayload:
        ...

    async def authenticate_bootstrap(
        self, req: BootstrapAuthRequest
    ) -> BootstrapAuthResult:
        ...

    async def get_bootstrap_stage(
        self, entity_type: str, entity_id: str
    ) -> BootstrapStage:
        ...

    async def issue_downstream_grant(
        self, req: DownstreamGrantRequest
    ) -> DownstreamAccessGrant:
        ...

    async def refresh_by_user_session(
        self, req: TokenRefreshRequest
    ) -> TokenBundle | None:
        ...

    async def verify_token(
        self, req: TokenVerifyRequest
    ) -> TokenVerificationResult | None:
        ...

    async def revoke_token(self, req: TokenRevokeRequest) -> None:
        ...

    async def validate_session(
        self, req: SessionValidateRequest
    ) -> Session | None:
        ...

File: gateway/src/interfaces/auth/auth_gateway_authority.go
package auth

import (
	"context"
	authmodel "gateway/src/models/auth"
)

// IAuthGatewayAuthorityClient 定义网关到认证中心统一鉴权门面的远端调用接口。
// Gateway 仅做转发，不在本地管理会话/令牌/挑战状态。
type IAuthGatewayAuthorityClient interface {
	InitChallenge(ctx context.Context, req *authmodel.ChallengeRequest) (*authmodel.ChallengePayload, error)
	AuthenticateBootstrap(ctx context.Context, req *authmodel.BootstrapAuthRequest) (*authmodel.BootstrapAuthResult, error)
	GetBootstrapStage(ctx context.Context, entityType authmodel.EntityType, entityID string) (authmodel.BootstrapStage, error)

	AuthenticateByPassword(ctx context.Context, req *authmodel.UserPasswordAuthRequest) (*authmodel.UserPasswordAuthResult, error)
	RefreshByUserSession(ctx context.Context, req *authmodel.TokenRefreshRequest) (*authmodel.TokenBundle, error)
	VerifyToken(ctx context.Context, req *authmodel.TokenVerifyRequest) (*authmodel.TokenVerificationResult, error)
	RevokeToken(ctx context.Context, req *authmodel.TokenRevokeRequest) error
	RevokeUserSession(ctx context.Context, req *authmodel.SessionRevokeRequest) error
	ValidateSession(ctx context.Context, req *authmodel.SessionValidateRequest) (*authmodel.Session, error)

	IssueDownstreamGrant(ctx context.Context, req *authmodel.DownstreamGrantRequest) (*authmodel.DownstreamAccessGrant, error)
}
